### PR TITLE
Add Custom Time to Lock/Warning

### PR DESCRIPTION
**Pull request by ragurney**
@zendesk/devex 
@jonmoter @grosser 

screen shot 2018-02-15 at 1 47 25 pm
screen shot 2018-02-15 at 1 47 35 pm
screen shot 2018-02-15 at 1 47 47 pm

### CODE REVIEW

**Comment by grosser** [comment]
check how much page-weight this adds

File: app/assets/javascripts/stages.js [hunk 1]
```diff
@@ -49,11 +49,18 @@ $(function() {
     });
   }
 
-  $(".want-lock").not(":disabled").click(function() {
-    var $form = $(this).next();
-    $(this).toggleClass("active");
-    $form.toggleClass("active");
-    $form.find(".lock-description").select();
+  // Datetime picker initialization
+  $(function () {
```
Comment (grosser) [inline app/assets/javascripts/stages.js]: can remove this, all this is already in a page-ready funtion

Comment (grosser) [reply app/assets/javascripts/stages.js]: also this is not only for stages, add a custom `locks.js` file

File: app/controllers/locks_controller.rb [hunk 2]
```diff
@@ -11,14 +11,18 @@ def index
...
+    if new_lock.valid?
+      new_lock.save!
+      respond_to do |format|
+        format.html { redirect_back notice: 'Locked', fallback_location: root_path }
+        format.json { render json: {lock: new_lock} }
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_back flash: { error: format_errors(new_lock) }, fallback_location: root_path }
+        format.json { head :unprocessable_entity }
```
Comment (grosser) [inline app/controllers/locks_controller.rb]: json errors are supposed to be 422 with error details

Comment (grosser) [reply app/controllers/locks_controller.rb]: can we check the date client-side instead ?

File: app/models/lock.rb [hunk 3]
```diff
@@ -120,4 +121,8 @@ def nil_out_blank_resource_type
   def unique_global_lock
     errors.add(:resource_id, :invalid) if global? && Lock.global.first
   end
+
+  def valid_delete_at
+    errors.add(:delete_at, 'Date must be in the future') unless delete_at.nil? || delete_at.future?
```
Comment (grosser) [inline app/models/lock.rb]: prefer if for multiple conditions `if !delete_at&.future?`

File: app/views/locks/_lock_button.html.erb [hunk 4]
```diff
@@ -1,9 +1,17 @@
...
+        <%= f.text_field :description, class: "lock-input", placeholder: "Description", required: warning %>
+        <span class="datetimepicker-container">
+          <%= f.text_field :delete_at, placeholder: "Expiration date", class: 'datetimepicker lock-input' %>
+        </span>
+      <%= f.submit text, class: "do-lock btn btn-warning lock-input" %>
+    <% end %>
+  </span>
+</span>
+
+
```
Comment (grosser) [inline app/views/locks/_lock_button.html.erb]: extra newline

**Comment by grosser** [comment]
how hard would it be to add a `1h / 2h / 4h / 1d` button ?

**Comment by ragurney** [comment]
Looks like page weight went from 3.1MB -> 3.2MB. I don't think it would be too hard to add the `1h / 2h / 4h / 1d` buttons... I could play around with it. Should just be some simple js to replace the value of the input field.

File: app/models/lock.rb (hunk 3 above)
Comment (ragurney) [reply app/models/lock.rb]: Rubocop seems to prefer the `unless` syntax here

**Comment by ragurney** [comment]
screen shot 2018-02-16 at 10 14 36 am

File: app/assets/javascripts/lock.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      }
+    });
+  }
+
+  function attachEventListeners() {
+    formatLockExpirationTime();
+    handleLockTimeButtons();
+  }
+
+  attachEventListeners();
```
Comment (grosser) [inline app/assets/javascripts/lock.js]: idk why we need 2 levels of indirection to attach an event :)

Comment (grosser) [reply app/assets/javascripts/lock.js]:
```js
// user clicks on pre-defined lock time buttons
$('.lock-times').click(function (e) {
```

File: app/assets/javascripts/lock.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        lockDeleteAtInput.val(lockExpireMoment.utc().format());
+      }
+    });
+  }
+
+  function handleLockTimeButtons() {
+    $('.lock-times').click(function (e) {
+      var buttonData = e.target.dataset;
+      var unit = buttonData.unit;
+      var quantity = parseInt(buttonData.num);
```
Comment (grosser) [inline app/assets/javascripts/lock.js]: parseInt(foo, 10)

File: app/assets/javascripts/lock.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      var lockDeleteAtInput = $(this).siblings('.datetimepicker-container').find('#lock_delete_at');
```
Comment (grosser) [inline app/assets/javascripts/lock.js]: having 2 different ways of getting the lock_delete_at seems a bit strange ...

maybe iterate over the $('.new-lock-form') ?

```js
$('.new-lock-form').each -> e
  delete_at = foo
  $(e).submit() stuff with delete_at
  $(e, .lock-times).click() stuff with delete_at
```

File: app/assets/javascripts/lock.js (hunk 5 above)
Comment (grosser) [reply app/assets/javascripts/lock.js]: more comments in general ... "JS for the lock form, users can x and Y"

File: app/assets/stylesheets/locks.scss [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  margin: .5em;
+}
+
+.lock-input {
+  border-radius: 4px;
+  border: 1px solid #ccc;
+  height: 34px;
+  padding: 6px;
+  vertical-align: bottom;
+  width: 405px;
```
Comment (grosser) [inline app/assets/stylesheets/locks.scss]: 100% did not work ?

File: app/controllers/locks_controller.rb [hunk 9]
```diff
@@ -11,14 +11,18 @@ def index
...
-    lock = Lock.create!(
-      params.require(:lock).
-        permit(:description, :resource_id, :resource_type, :warning, :delete_in).
-        merge(user: current_user)
-    )
-    respond_to do |format|
-      format.html { redirect_back notice: 'Locked', fallback_location: root_path }
-      format.json { render json: {lock: lock} }
+    new_lock = Lock.new(lock_params)
+    if new_lock.valid?
```
Comment (grosser) [inline app/controllers/locks_controller.rb]: same as `if new_lock.save`

File: app/controllers/locks_controller.rb [hunk 10]
```diff
@@ -11,14 +11,18 @@ def index
...
+      new_lock.save!
+      respond_to do |format|
+        format.html { redirect_back notice: 'Locked', fallback_location: root_path }
+        format.json { render json: {lock: new_lock} }
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_back flash: { error: format_errors(new_lock) }, fallback_location: root_path }
+        format.json { render json: { error: new_lock.errors.full_messages }, status: :unprocessable_entity }
```
Comment (grosser) [inline app/controllers/locks_controller.rb]: use json_error method for consistency

```
render_json_error 422, exception.record.errors
```

File: app/controllers/locks_controller.rb [hunk 11]
```diff
@@ -11,14 +11,18 @@ def index
...
   def create
-    lock = Lock.create!(
-      params.require(:lock).
-        permit(:description, :resource_id, :resource_type, :warning, :delete_in).
-        merge(user: current_user)
-    )
-    respond_to do |format|
-      format.html { redirect_back notice: 'Locked', fallback_location: root_path }
-      format.json { render json: {lock: lock} }
+    new_lock = Lock.new(lock_params)
```
Comment (grosser) [inline app/controllers/locks_controller.rb]: also new_lock -> lock ?

File: app/models/lock.rb (hunk 3 above)
Comment (grosser) [reply app/models/lock.rb]: how about `if delete_at && !delete_at.future?`

Comment (grosser) [reply app/models/lock.rb]: how about `if delete_at&.past?` 🎉

File: app/assets/stylesheets/locks.scss (hunk 8 above)
Comment (ragurney) [reply app/assets/stylesheets/locks.scss]: No it didn't.. I played around with it for a bit and for some reason it was ignoring my right margin, so I just went with hard-coded px. I could dig into it a little more though if you prefer 100%.

File: app/controllers/locks_controller.rb (hunk 11 above)
Comment (ragurney) [reply app/controllers/locks_controller.rb]: There's a method `lock` in the controller, and I didn't want to cause confusion by making the variable shadow the method, and `new_lock` is somewhat more descriptive?

**Comment by grosser** [comment]
meh ... doing that all the time in other controllers, but either way is
fine ...

File: app/assets/javascripts/lock.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  $('.datetimepicker').datetimepicker();
+
+  $('.new-lock-form').each(function() {
+    var form = $(this);
+    var lockDeleteAtInput = form.find('#lock_delete_at');
+    var lockExpirationTimeButtons = form.find('.lock-times');
+
+    formatLockExpirationTime(form, lockDeleteAtInput);
```
Comment (grosser) [inline app/assets/javascripts/lock.js]: could save a bit of passing objects around by doing:
`form.submit(function() { formatLockExpirationTime(lockDeleteAtInput) })`

**Pull request merged**